File: leetcode-python-go/733. 图像渲染/main.py
from typing import List
import collections

image = [[1,1,1],[1,1,0],[1,0,1]]
sr = 1
sc = 1
newColor = 2

def floodFill(image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
    currColor = image[sr][sc]
    if currColor == color:
        return image
    
    n, m = len(image), len(image[0])
    que = collections.deque([(sr, sc)])
    image[sr][sc] = color
    while que:
        x, y = que.popleft()
        for mx, my in [(x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1)]:
            if 0 <= mx < n and 0 <= my < m and image[mx][my] == currColor:
                que.append((mx, my))
                image[mx][my] = color
    
    return image

print(floodFill(image, sr, sc, newColor))

File: leetcode-python-go/733. 图像渲染/main.go
package main

import (
	"fmt"
)

var (
	dx = []int{1, 0, 0, -1}
	dy = []int{0, 1, -1, 0}
)

func floodFill(image [][]int, sr int, sc int, color int) [][]int {
	currColor := image[sr][sc]
	if currColor == color {
		return image
	}
	n, m := len(image), len(image[0])
	queue := [][]int{}
	queue = append(queue, []int{sr, sc})
	image[sr][sc] = color
	for i := 0; i < len(queue); i++ {
		cell := queue[i]
		for j := 0; j < 4; j++ {
			mx, my := cell[0]+dx[j], cell[1]+dy[j]
			if mx >= 0 && mx < n && my >= 0 && my < m && image[mx][my] == currColor {
				queue = append(queue, []int{mx, my})
				image[mx][my] = color
			}
		}
	}
	return image
}

func main() {
	image := [][]int{{1, 1, 1}, {1, 1, 0}, {1, 0, 1}}
	sr := 1
	sc := 1
	newColor := 2
	results := floodFill(image, sr, sc, newColor)
	fmt.Println(results)
}
